### DISCUSSION TITLE
Is BinaryRepresentation.TWOS_COMPLEMENT broken?

### DISCUSSION THREAD

**Discussion by darrenxonaspace** [cocotb/cocotb#3444] [Q&A] [answered]
I want to assign a negative 2's complement value to a 4-bit signal. Here is what I have tried.

```python
@cocotb.test()
async def twos_comp_test(dut):
  neg_int = -6
  bv = BinaryValue(n_bits=4, binaryRepresentation=BinaryRepresentation.TWOS_COMPLEMENT)
  bv.signed_integer = neg_int
  # bv.integer = neg_int
  # bv.value = neg_int
  # bv.assign(neg_int)
  print(f"neg_int: {neg_int}")
  print(f"bv: {bv}")
  print(f"bv.binstr: {bv.binstr}")
  print(f"bv.value {bv.value}")
  print(f"bv.integer: {bv.integer}")
  print(f"bv.signed_integer: {bv.signed_integer}")
```
The test was repeated with all of the commented-out alternative assignment strategies for `bv`, but the result was the same every time. The return binary is `0101`, or +5, rather than `1010` as expected.

```bash
neg_int: -6
bv: 0101
bv.binstr: 0101
bv.value 5
bv.integer: 5
bv.signed_integer: 5
```

Am I missing something about how to use this? Is there a different way to have cocotb use two's complement when passing a signal to/from the DUT?

**Comment by ktbarrett** [comment, marked as answer]
BinaryValue is known to be broken and will be replaced in the next release cycle. Use the `LogicArray` in `cocotb.types`.

**Reply by darrenxonaspace** [reply]
@ktbarrett thank you! This should really be updated in the documentation. I was pointed directly to `BinaryValue` for signed numbers by [the docs here](https://docs.cocotb.org/en/stable/writing_testbenches.html), and there is no mention either there or on the `binary` library page of it being broken.

![image](https://github.com/cocotb/cocotb/assets/147874182/b3617fbd-a6a4-465e-bfe1-7603ecdadf42)

**Reply by darrenxonaspace** [reply]
Perhaps I am looking in the wrong place for docs, but your comment is the first place I saw `LogicArray` mentioned, which seems like a very useful analog to `std_logic_vector`

**Reply by ktbarrett** [reply]
We probably should have done a better job calling attention to it before we released 1.8. But seeing as there will not be another release before 2.0, where the change will occur, there isn't much to do.

**Reply by imphil** [reply]
To add a bit more nuance to that: For most use cases and users, `BinaryValue` works just fine (otherwise it wouldn't have been around for so long). At the same time, `BinaryValue` has deficiencies in its design, especially around the handling signed vs unsigned data. We cannot fix those in cocotb 1.x due to backwards-compat concerns. 

When using `LogicArray` and friends in cocotb 1.x, keep in mind that they are experimental and their interface might change.

If you're interested in that topic, please have a look at https://github.com/cocotb/cocotb/pull/3244 and related PRs and issues, where we're trying to come up with a next-generation design for this area.

**Reply by darrenxonaspace** [reply]
@imphil very interesting read, and exciting to see some of what's being worked on! Still, it seems that the current docs really should not have an entire "Signed and unsigned values" section that not only fails to mention the known issues with BinaryValue, but explicitly directs users to functionality that is known to be broken.

Would it make sense for me to make a small edit and pull request, or is there a reason that updating the docs before the 2.0 release is not possible?

**Reply by imphil** [reply]
The docs always match a release, and we currently have no strong reason to do a 1.8.2 release -- but I wouldn't be surprised if find a reason to do such a release in the not too distant future, so: yes, you are invited to do a pull request with the doc fix. Please do it for against the *master* branch, and I'll then cherry-pick it into the stable/1.8 branch. I cannot promise any date when we'll do the next release, but I'll keep it on my radar.
